ziglang

<!--
IMPORTANT: we are not accepting issues about Termux
installations running on Android 5.x-6.x.
-->

**Package description**
Zig is a general-purpose programming language designed for robustness, optimality, and maintainability.
<!--
Describe what this package do and why it should be
available. Make sure that requested package supports
cross-compilation and can run on AArch64 and ARM.
-->

**Link to home page and sources**
https://github.com/ziglang/zig

**Additional information**
It compiles fine using the embedded lld but in the linking step it gets struck on a atomic error.

i've tried building zig yesterday on device, but it fails on linking. here's what i did so far:
* applied this little patch to zig repo:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 467918557..fb80a93d1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -448,6 +448,8 @@ set_target_properties(zig0 PROPERTIES
     LINK_FLAGS ${EXE_LDFLAGS}
 )
 target_link_libraries(zig0 zigcompiler)
+target_link_libraries(zig0 atomic)
+target_link_libraries(zig0 android-spawn)
 
 if(MSVC)
     set(LIBSTAGE2 "${CMAKE_BINARY_DIR}/zigstage2.lib")
```
* after running `cmake ..`, manually changed in CMakeCache.txt line
```
ZIG_PREFER_CLANG_CPP_DYLIB:BOOL=off
```
to
```
ZIG_PREFER_CLANG_CPP_DYLIB:BOOL=on
```
after doing all of above, build fails with
```
[ 97%] Linking CXX executable zig0
/data/data/com.termux/files/usr/bin/arm-linux-androideabi-ld: zig_cpp/libzig_cpp.a(zig_clang_cc1_main.cpp.o): in function `long std::__ndk1::__libcpp_atomic_refcount_increment<long>(long&)':
/data/data/com.termux/files/usr/include/c++/v1/memory:3350: undefined reference to `__atomic_fetch_add_4'
/data/data/com.termux/files/usr/bin/arm-linux-androideabi-ld: zig_cpp/libzig_cpp.a(zig_clang_cc1_main.cpp.o): in function `long std::__ndk1::__libcpp_atomic_refcount_decrement<long>(long&)':
/data/data/com.termux/files/usr/include/c++/v1/memory:3361: undefined reference to `__atomic_fetch_add_4'
/data/data/com.termux/files/usr/bin/arm-linux-androideabi-ld: zig_cpp/libzig_cpp.a(zig_clang_driver.cpp.o): in function `int std::__ndk1::__cxx_atomic_fetch_sub<int>(std::__ndk1::__cxx_atomic_base_impl<int>*, int, std::__ndk1::memory_order)':
/data/data/com.termux/files/usr/include/c++/v1/atomic:1036: undefined reference to `__atomic_fetch_sub_4'
/data/data/com.termux/files/usr/bin/arm-linux-androideabi-ld: zig_cpp/libzig_cpp.a(zig_clang_driver.cpp.o): in function `int std::__ndk1::__cxx_atomic_fetch_add<int>(std::__ndk1::__cxx_atomic_base_impl<int>*, int, std::__ndk1::memory_order)':
/data/data/com.termux/files/usr/include/c++/v1/atomic:1014: undefined reference to `__atomic_fetch_add_4'
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/zig0.dir/build.make:135: zig0] Error 1
make[1]: *** [CMakeFiles/Makefile2:225: CMakeFiles/zig0.dir/all] Error 2
make: *** [Makefile:150: all] Error 2
```
which should have been solved by
```
+target_link_libraries(zig0 atomic)
```
but it still happens nonetheless :(

--- a/build-package.sh
+++ b/build-package.sh
@@ -67,6 +67,10 @@ source "$TERMUX_SCRIPTDIR/scripts/build/setup/termux_setup_golang.sh"
 # shellcheck source=scripts/build/setup/termux_setup_rust.sh
 source "$TERMUX_SCRIPTDIR/scripts/build/setup/termux_setup_rust.sh"
 
+# Utility function for zig-using packages to setup a zig toolchain.
+# shellcheck source=scripts/build/setup/termux_setup_zig.sh
+source "$TERMUX_SCRIPTDIR/scripts/build/setup/termux_setup_zig.sh"
+
 # Utility function to setup a current ninja build system.
 # shellcheck source=scripts/build/setup/termux_setup_ninja.sh
 source "$TERMUX_SCRIPTDIR/scripts/build/setup/termux_setup_ninja.sh"

--- a/packages/zig/build.sh
+++ b/packages/zig/build.sh
@@ -1,0 +1,23 @@
+TERMUX_PKG_HOMEPAGE=https://ziglang.org
+TERMUX_PKG_DESCRIPTION="General-purpose programming language and toolchain for maintaining robust, optimal, and reusable software."
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_MAINTAINER="@leapofazzam123"
+TERMUX_PKG_VERSION=0.9.0
+TERMUX_PKG_SRCURL=https://ziglang.org/download/$TERMUX_PKG_VERSION/zig-$TERMUX_PKG_VERSION.tar.xz
+TERMUX_PKG_SHA256=cd1be83b12f8269cc5965e59877b49fdd8fa638efb6995ac61eb4cea36a2e381
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_DEPENDS="libandroid-spawn, libc++, libxml2, ncurses, zlib"
+TERMUX_PKG_BUILD_DEPENDS="llvm, libllvm-static"
+TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
+-DZIG_PREFER_CLANG_CPP_DYLIB=OFF
+-DLLVM_LIBDIRS=$TERMUX_PREFIX/lib
+"
+
+termux_step_pre_configure() {
+	termux_setup_zig
+	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+="
+		-DZIG_EXECUTABLE=$(which zig)
+		-DZIG_TARGET_TRIPLE=$ZIG_TARGET_NAME
+		"
+	LDFLAGS+=" -landroid-spawn -lncursesw -lxml2 -lz"
+}

--- a/packages/zig/cmake-Findllvm.cmake.patch
+++ b/packages/zig/cmake-Findllvm.cmake.patch
@@ -1,0 +1,11 @@
+--- a/cmake/Findllvm.cmake
++++ b/cmake/Findllvm.cmake
+@@ -353,6 +353,8 @@
+   FIND_AND_ADD_LLVM_LIB(LLVMBinaryFormat)
+   FIND_AND_ADD_LLVM_LIB(LLVMSupport)
+   FIND_AND_ADD_LLVM_LIB(LLVMDemangle)
++  FIND_AND_ADD_LLVM_LIB(Polly)
++  FIND_AND_ADD_LLVM_LIB(PollyISL)
+ endif()
+ 
+ include(FindPackageHandleStandardArgs)

--- a/scripts/build/setup/termux_setup_zig.sh
+++ b/scripts/build/setup/termux_setup_zig.sh
@@ -1,0 +1,44 @@
+termux_setup_zig() {
+	local ZIG_VERSION=0.9.0
+	local ZIG_FOLDER
+
+	if [ "${TERMUX_PACKAGES_OFFLINE-false}" = "true" ]; then
+		ZIG_FOLDER=${TERMUX_SCRIPTDIR}/build-tools/zig-${ZIG_VERSION}
+	else
+		ZIG_FOLDER=${TERMUX_COMMON_CACHEDIR}/zig-${ZIG_VERSION}
+	fi
+
+	if [ "$TERMUX_ON_DEVICE_BUILD" = "false" ]; then
+		if [ ! -x "$ZIG_FOLDER/zig" ]; then
+			mkdir -p "$ZIG_FOLDER"
+			local ZIG_TARBALL=$TERMUX_PKG_TMPDIR/zig-$ZIG_VERSION.zip
+			termux_download https://ziglang.org/download/$ZIG_VERSION/zig-linux-x86_64-$ZIG_VERSION.tar.xz \
+				"$ZIG_TARBALL" \
+				5c55344a877d557fb1b28939785474eb7f4f2f327aab55293998f501f7869fa6
+			tar xf "$ZIG_TARBALL" -C "$ZIG_FOLDER" --strip-components=1
+		fi
+		export PATH=$ZIG_FOLDER:$PATH
+	else
+		local ZIG_PKG_VERSION=$(bash -c ". $TERMUX_SCRIPTDIR/packages/zig/build.sh; echo \$TERMUX_PKG_VERSION")
+		if ([ ! -e "$TERMUX_BUILT_PACKAGES_DIRECTORY/zig" ] ||
+		    [ "$(cat "$TERMUX_BUILT_PACKAGES_DIRECTORY/zig")" != "$ZIG_PKG_VERSION" ]) &&
+		   [ "$(dpkg-query -W -f '${db:Status-Status}\n' zig 2>/dev/null)" != "installed" ]; then
+			echo "Package 'zig' is not installed."
+			echo "You can install it with"
+			echo
+			echo "  pkg install zig"
+			echo
+			echo "or build it from source with"
+			echo
+			echo "  ./build-package.sh zig"
+			echo
+			exit 1
+		fi
+	fi
+
+	if [ $TERMUX_ARCH = "i686" ]; then
+		ZIG_TARGET_NAME=i386-linux-android
+	else
+		ZIG_TARGET_NAME=$TERMUX_ARCH-linux-android
+	fi
+}

--- a/scripts/setup-offline-bundle.sh
+++ b/scripts/setup-offline-bundle.sh
@@ -47,6 +47,9 @@ mkdir -p "$TERMUX_PKG_TMPDIR"
 #(. "$TERMUX_SCRIPTDIR"/scripts/build/setup/termux_setup_rust.sh
 #	termux_setup_rust
 #)
+(. "$TERMUX_SCRIPTDIR"/scripts/build/setup/termux_setup_zig.sh
+	termux_setup_zig
+)
 rm -rf "${TERMUX_PKG_TMPDIR}"
 (test -d "$TERMUX_SCRIPTDIR"/build-tools/android-sdk && test -d "$TERMUX_SCRIPTDIR"/build-tools/android-ndk && exit 0
 	"$TERMUX_SCRIPTDIR"/scripts/setup-android-sdk.sh

